gsoc2020/20200303-health-check-wjayesh.md

```diff
@@ -0,0 +1,254 @@
+ 
+""" 
+# Args: 
+# indicators: A set of true/false values for the input resources 
+#  false indicating bad health 
+# Returns: 
+#  Status report indicating if operations successful 
+} 
+```
+
+```python
+deployAndExposePod() { 
+"""  
+Method to create lightweight pod inside the cluster with purpose of 
+accessing the Kubernetes networking and running `dig` on the ClusterIP.	
+""" 
+# Args: 
+# None 
+ 
+# Returns: 
+#  IP address of the service, pod name and other information 
+} 
+```
+ 
+
+Timeline 
+-
+
+**Before 27 April:** 
+
+I’ll get comfortable with the Go language and DNS concepts. 
+
+
+**27 April to 18 May:** 
+ 
+I’ll bond with the community and understand what test systems, coding practices and principles are followed. I’ll stay active on community channels and discuss the idea and solution further to finalize and perfect the approach. 
+
+
+**18 May to 31st May:** 
+ 
+Develop code to enable the application to determine what scenario and environment it is deployed in.  
+I will implement a priority order for the different approaches when two or more are existing simultaneously; after deliberations with the mentors.  
+Ultimately, the app at this stage should be able to transfer control to one of the four cases. 
+ 
+**1st June to 7th June:** 
+ 
+Test the existing code rigorously and document the process. 
+ 
+**7th June to 21st June:** 
+ 
+Take on the first case (deployed as a pod) and implement part of the specified API that caters to this situation. A lot of the code written in this period could be reused with some additions or logic control for the remaining cases.  
+ 
+**22nd June to 30th June:** 
+ 
+Test and document the existing code thoroughly.  
+ 
+ 
+ 
+**1st July to 7th July:** 
+ 
+Implement the part of the API specific to the second case. Some code will need to be added to get the `NodePort`/`LoadBalancer` IP in this case. Will also look for other modifications required.  
+ 
+**7th July to 14th July:** 
+ 
+Test and document the application up to this stage and ensure the two cases report correct results and function as expected.  
+ 
+**14th July to 21st July:** 
+ 
+Implement the deployAndExposePod function and check compatibility with code in the existing functions. Figure out if any modifications or utilities needed. 
+ 
+**21st July to 28th July:** 
+ 
+Test and document the code and ensure all the three cases work as expected.  
+ 
+**28th July to 10th August:** 
+ 
+Explore options for solving the fourth case and look for any hacks or contributions required to make it work.  
+Will also further test the code in different conditions and better the documentation of the entire project. 
+ 
+**10th to 17th August:** 
+Buffer to accommodate for unexpected delays or roadblocks.    
+ 
+ 
+
+ 
+
+## Questions and Discussion Topics
+
+- Are there any deployment cases that I have missed?
+- Are there any modifications to the strategies listed above, that should be made?
+ 
+ 
+
```

Comment:
@WJayesh , thank you for the proposal! I am waiting for your questions and modifications.

Reply:
Thank you for the suggestions :) 
I'll start with the modifications right away!

Reply:
@WJayesh, print your proposal. Then, read it and correct it. Rewrite. Repeat 5 times. That’s your draft 2 ;-) after draft 5, you will have the opportunity to dive into the wonderful world of miekg/dns, because that is the “core” of CoreDNS.

Reply:
I understand that the implementation section needs changes to include the use of the dns library for tasks like sending UDP queries. I will read meikg/dns for it. 

It is, however, not clear to me what the last line of your comment meant. As I understand it, the work done on this tool won't be affecting the CoreDNS binary as it will be an independent application. I cannot connect the use of meikg/dns in CoreDNS to its use in the health check binary.

Reply:
>  I cannot connect the use of meikg/dns in CoreDNS to its use in the health check binary.

@WJayesh, you can use `meikg/dns` to make DNS requests. I saw you are mentioning `dig`. So, no `dig` but `meikg/dns`.